server/src/main/java/org/elasticsearch/cluster/coordination/StableMasterHealthIndicatorService.java

```diff
@@ -0,0 +1,449 @@
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.health.ServerHealthComponents.CLUSTER_COORDINATION;
+
+/**
+ * This indicator reports the health of master stability.
+ * If we have had a master within the last 30 seconds, and that master has not changed more than 3 times in the last 30 minutes, then
+ * this will report GREEN.
+ * If we have had a master within the last 30 seconds, but that master has changed more than 3 times in the last 30 minutes (and that is
+ * confirmed by checking with the last-known master), then this will report YELLOW.
+ * If we have not had a master within the last 30 seconds, then this will will report RED with one exception. That exception is when:
+ * (1) no node is elected master, (2) this node is not master eligible, (3) some node is master eligible, (4) we ask a master-eligible node
+ * to run this indicator, and (5) it comes back with a result that is not RED.
+ * Since this indicator needs to be able to run when there is no master at all, it does not depend on the dedicated health node (which
+ * requires the existence of a master).
+ */
+public class StableMasterHealthIndicatorService implements HealthIndicatorService, ClusterStateListener {
+
+    public static final String NAME = "master_is_stable";
+    private static final String HELP_URL = "https://ela.st/fix-master";
+
+    private final ClusterService clusterService;
+    private final MasterHistoryService masterHistoryService;
+    /**
+     * This is the amount of time we use to make the initial decision -- have we seen a master node in the very recent past?
+     */
+    private final TimeValue veryRecentPast;
+    /**
+     * If the master transitions from a non-null master to a null master at least this many times it starts impacting the health status.
+     */
+    private final int unacceptableNullTransitions;
+    /**
+     * If the master transitions from one non-null master to a different non-null master at least this many times it starts impacting the
+     * health status.
+     */
+    private final int unacceptableIdentityChanges;
+
+    private static final Logger logger = LogManager.getLogger(StableMasterHealthIndicatorService.class);
+
+    // This is the default amount of time we look back to see if we have had a master at all, before moving on with other checks
+    private static final TimeValue DEFAULT_VERY_RECENT_PAST = new TimeValue(30, TimeUnit.SECONDS);
+    private static final TimeValue SMALLEST_ALLOWED_VERY_RECENT_PAST = new TimeValue(1, TimeUnit.SECONDS);
+
+    // This is the default number of times that it is OK to have a master go null. Any more than this will be reported as a problem
+    private static final int DEFAULT_ACCEPTABLE_NULL_TRANSITIONS = 3;
```

Comment:
I don't think we need all of these as private static variables, we can put the comment and the actual value (3 in this case) in the setting itself, which is already public and static?

I prefer to have the hard value with the setting itself, so if you look at the `Setting` you don't have to do another reference jump to see the default value.

Reply:
Done.